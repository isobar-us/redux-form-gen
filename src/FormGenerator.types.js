import type {FieldsType, CustomFieldTypes} from './types';
import type {CustomOperators} from './conditionalUtils.types';
import type {LookupTable} from './utils.types';

export type Props = {
  fields: FieldsType,
  customQuestionProps?: {
    [key: string]: Object
  },
  visibleDepth?: string,
  customFieldTypes?: CustomFieldTypes,
  customOperators?: CustomOperators,
  display?: 'stacked' | 'inline',
  disabled?: boolean,
  lookupTable: LookupTable
};

export type State = {
  wasGenerated: true,
  lookupTable: Object | null
};
