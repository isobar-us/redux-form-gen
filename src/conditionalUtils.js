// @flow
import get from 'lodash/get';
import has from 'lodash/has';
import hasIn from 'lodash/hasIn';
import includes from 'lodash/includes';
import omit from 'lodash/omit';
import isEqual from 'lodash/isEqual';
import isNumber from 'lodash/isNumber';
import isNil from 'lodash/isNil';
import isString from 'lodash/isString';
import {isNilOrEmpty, isFieldFilled, isFieldValid} from './validators';

import type {FieldType} from './types';
import type {ConditionalOperators, EvalCondOptions, ConditionalObject} from './conditionalUtils.types';

/*
  Helpers
 */

// used for render lifecycle where we don't care about conditionalValid
export const isCondField = (field: FieldType) =>
  has(field, 'conditionalVisible') || has(field, 'conditionalRequired') || has(field, 'conditionalDisabled');

const getCondValueKey = ({cond, valueKey}) => (has(cond, 'questionId') ? cond.questionId : valueKey);

const getOperator = (options, key) => {
  const {customOperators} = options;
  const operators = {...ops, ...(customOperators || {})};
  const op = get(operators, key);
  if (isNil(op)) {
    console.error(`[FormGenerator] Unknown conditional operator "${key}". Condition:`, options.cond);
  }
  return op;
};

/*
  Conditional Operators
 */
const ops: ConditionalOperators = {
  equals: ({value, param}) => isEqual(value, param),
  // cond
  and: ({value, param, ...options}) =>
    param.reduce((and, subCond) => and && evalCond({...options, cond: subCond}), true),
  or: ({value, param, ...options}) => param.reduce((or, subCond) => or || evalCond({...options, cond: subCond}), false),
  not: ({value, param, ...options}) => !evalCond({...options, cond: param}),
  cond: ({value, param, ...options}) => evalCond({...options, cond: param}),
  compare: (options) => {
    const {value, param, data} = options;

    const conds = Object.keys(omit(param, ['questionId']));
    return conds.length > 0
      ? conds.reduce((result, key) => {
          // will AND all the cond props
          const operator = getOperator(options, key);
          if (isNil(operator)) {
            return true;
          }
          const remoteQuestionId = get(param, key);
          const compareParam = get(
            data,
            getCondValueKey({
              ...options,
              cond: {
                ...(remoteQuestionId && {questionId: remoteQuestionId})
              }
            })
          );
          return result && operator({...options, value, param: compareParam});
        }, true)
      : true;
  },
  // generator hooks
  filled: ({value, param, ...options}) => {
    let fieldFilled = true;
    if (has(options, 'lookupTable')) {
      const {lookupTable} = options;
      const field = get(lookupTable, getCondValueKey(options));
      if (!isNil(field)) {
        // if field is defined in the lookupTable, run isFieldFilled()
        fieldFilled = isFieldFilled({...options, field, fieldOptions: null});
      } else {
        // if field is not defined in the lookupTable, just fall back to isNilOrEmpty
        fieldFilled = !isNilOrEmpty(value);
      }
    } else {
      console.warn('[Form Generator] attempted to check `filled` without a lookupTable. Condition:', options.cond); // TODO finalise this warning
      fieldFilled = !isNilOrEmpty(value);
    }
    return param === true ? fieldFilled : !fieldFilled;
  },
  // TODO experimental operator to check if this or another field is valid...
  // valid: ({value, param, ...options}) => {
  //   const {cond, lookupTable} = options;
  //   let field = options.field;
  //
  //   if (has(cond, 'questionId')) {
  //     if (has(options, 'lookupTable')) {
  //       field = get(lookupTable, cond.questionId);
  //     } else {
  //       console.warn('[Form Generator] attempted to check `valid` without a lookupTable. Condition:', options.cond); // TODO finalise this warning
  //     }
  //   }
  //
  //   const fieldValid = isFieldValid({
  //     ...options,
  //     field,
  //     fieldOptions: null,
  //   });
  //   return param === true ? fieldValid : !fieldValid;
  // },
  // lodash
  includes: ({value, param}) => includes(value, param),
  // comparison
  greaterThan: ({value, param}) => {
    if (isNilOrEmpty(value)) {
      return false;
    }
    const parsedValue = isNumber(value) ? value : parseFloat(value);
    return parsedValue > param;
  },
  lessThan: ({value, param}) => {
    if (isNilOrEmpty(value)) {
      return false;
    }
    const parsedValue = isNumber(value) ? value : parseFloat(value);
    return parsedValue < param;
  },
  greaterThanEqual: ({value, param}) => {
    if (isNilOrEmpty(value)) {
      return false;
    }
    const parsedValue = isNumber(value) ? value : parseFloat(value);
    return parsedValue >= param;
  },
  lessThanEqual: ({value, param}) => {
    if (isNilOrEmpty(value)) {
      return false;
    }
    const parsedValue = isNumber(value) ? value : parseFloat(value);
    return parsedValue <= param;
  },
  // length
  length: ({value, param}) => (isNil(value) ? false : hasIn(value, 'length') ? value.length === param : true),
  minLength: ({value, param}) => (isNil(value) ? false : hasIn(value, 'length') ? value.length >= param : true),
  maxLength: ({value, param}) => (hasIn(value, 'length') ? value.length <= param : true),
  // regex
  pobox: ({value, param}) => (param && isString(value) ? /p.*o.* box/gi.test(value) : true),
  email: ({value, param}) => (param && isString(value) ? /\S+@\S+\.\S+/.test(value) : true),
  regex: ({value, param}) => (isString(value) ? RegExp(param).test(value) : true)
};

/*
  Conditional Operator Aliases
 */
ops.gt = ops.greaterThan;
ops.lt = ops.lessThan;
ops.min = ops.greaterThanEqual;
ops.max = ops.lessThanEqual;
ops.value = ops.equals;

const defaultElseHandler = (options) => ops.filled({...options, param: true});

const defaultValueKey = '_value';

export const evalCond = (options: EvalCondOptions) => {
  options = {
    elseHandler: defaultElseHandler,
    valueKey: defaultValueKey,
    ...options
  };
  const {cond, data, elseHandler} = options;

  // TODO add support for getFieldPath() when doing get(data, path) ?
  const value = get(data, getCondValueKey(options));
  const conds = Object.keys(omit(cond, 'questionId'));

  return conds.length > 0
    ? conds.reduce((result, key) => {
        // will AND all the cond props
        const operator = getOperator(options, key);
        if (isNil(operator)) {
          return true;
        }
        const param = get(cond, key);
        return result && operator({...options, value, param});
      }, true)
    : elseHandler({...options, value});
};

const condValidElseHandler = (options) => {
  const {cond, lookupTable} = options;
  let field = options.field;
  let fieldOptions = null;

  if (has(cond, 'questionId')) {
    if (has(options, 'lookupTable')) {
      field = get(lookupTable, cond.questionId);
      fieldOptions = null;
    } else {
      console.warn('[Form Generator] attempted to check `valid` without a lookupTable. Condition:', options.cond); // TODO finalise this warning
    }
  }

  return isFieldValid({
    ...options,
    field,
    fieldOptions
  });

  // TODO hook into experimental `valid` operator when ready
  // return ops.valid({...options, param: true});
};
export const evalCondValid = (options: EvalCondOptions) => {
  const {data} = options;
  const value = get(data, getCondValueKey(options));

  // can not validate if the field is empty, so return true
  if (isNil(value) || value === '') {
    return true;
  }

  return evalCond({
    elseHandler: condValidElseHandler,
    ...options
  });
};

export default {
  evalCond,
  evalCondValid
};

// TODO consider renaming to getCondDependentFieldNames or something similar
export const condDependentFields = (cond: ConditionalObject) => {
  if (has(cond, 'questionId')) {
    return [cond.questionId];
  } else if (has(cond, 'and') && Array.isArray(cond.and)) {
    return cond.and.reduce((a, subCond) => a.concat(condDependentFields(subCond)), []);
  } else if (has(cond, 'or') && Array.isArray(cond.or)) {
    return cond.or.reduce((a, subCond) => a.concat(condDependentFields(subCond)), []);
  } else if (has(cond, 'not') && cond.not) {
    return condDependentFields(cond.not);
  }
  return [];
};
