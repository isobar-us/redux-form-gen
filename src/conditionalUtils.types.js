import type {FieldType} from './types';
export type ConditionalObject = {
  questionId: string,
  equals?: mixed,
  // cond
  and?: Array<ConditionalObject>,
  or?: Array<ConditionalObject>,
  not?: ConditionalObject,
  // lodash
  filled?: boolean,
  includes?: mixed | Array<mixed>,
  // comparison
  greaterThan?: number,
  lessThan?: number,
  greaterThanEqual?: number,
  lessThanEqual?: number,
  // length
  length: number,
  minLength: number,
  maxLength: number,
  // regex
  pobox: boolean,
  email: boolean,
  regex: boolean
};

export type ConditionalDependency = {
  questionId: string,
  globalScope?: Boolean
};

export type FieldDependencyOptions = {
  field: FieldType,
  parentPath: String
};

export type EvalCondOptions = {
  elseHandler?: Function,
  valueKey?: string,
  cond: ConditionalObject,
  customOperators: CustomOperators // eslint-disable-line no-use-before-define
};

export type CustomOperators = {
  [key: string]: {(options: EvalCondOptions): boolean}
};

export type ConditionalOperatorOptions = {
  value: mixed,
  param: mixed,
  ...EvalCondOptions
};

type ConditionalOperatorNumberParam = {
  (options: {...ConditionalOperatorOptions, value: number, param: number}): boolean
};
type ConditionalOperatorObjectParam = {
  (options: {...ConditionalOperatorOptions, value: Object, param: boolean}): boolean
};

type ConditionalOperator = {(options: ConditionalOperatorOptions): boolean};

export type ConditionalOperators = {
  equals: ConditionalOperator,
  and: {(options: {...ConditionalOperatorOptions, param: Array<ConditionalObject>}): boolean},
  or: {(options: {...ConditionalOperatorOptions, param: Array<ConditionalObject>}): boolean},
  not: {(options: {...ConditionalOperatorOptions, param: ConditionalObject}): boolean},
  filled: ConditionalOperator,
  includes: {
    (options: {...ConditionalOperatorOptions, value: Array<mixed>, param: mixed}): boolean
  },
  greaterThan: ConditionalOperatorNumberParam,
  lessThan: ConditionalOperatorNumberParam,
  greaterThanEqual: ConditionalOperatorNumberParam,
  lessThanEqual: ConditionalOperatorNumberParam,
  length: ConditionalOperatorObjectParam,

  gt?: ConditionalOperatorNumberParam,
  lt?: ConditionalOperatorNumberParam,
  min?: ConditionalOperatorNumberParam,
  max?: ConditionalOperatorNumberParam,
  value?: ConditionalOperator
};
