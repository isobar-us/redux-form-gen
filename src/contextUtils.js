// @flow
import React from 'react';
import PropTypes from 'prop-types';

import createReactContext from 'create-react-context';
export const GenContext = createReactContext({wasGenerated: false});

export const consumeGenContext = (Component: Function) => (props: mixed) => (
  <GenContext.Consumer>
    {(gen) => <Component {...props} gen={gen} /> }
  </GenContext.Consumer>
);

// inspired by
// https://medium.com/react-ecosystem/how-to-handle-react-context-a7592dfdcbc

// const provideContext = (childContextTypes, getChildContext) => (WrappedComponent) => {
//   class ContextProvider extends Component {
//     static childContextTypes = childContextTypes;
//     getChildContext = () => getChildContext(this.props);
//
//     render() {
//       return <WrappedComponent {...this.props} />;
//     }
//   }
//   return ContextProvider;
// };

const consumeContext = (contextTypes) => (Component: Function) => {
  /* The context is passed as props. This way the component is
   completely decoupled from the context API.
  */
  const ContextConsumer = (props: mixed, context: mixed) => <Component {...props} {...context} />;
  ContextConsumer.contextTypes = contextTypes;
  return ContextConsumer;
};

export const reduxFormContext = {
  _reduxForm: PropTypes.object
};

export const consumeReduxFormContext = consumeContext(reduxFormContext);
