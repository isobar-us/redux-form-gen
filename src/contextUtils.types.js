import type {Props as FormGeneratorProps} from './FormGenerator.types';

export type GenContextProps = {
  getCachedValue: Function,
  setCachedValue: Function
} & FormGeneratorProps;
