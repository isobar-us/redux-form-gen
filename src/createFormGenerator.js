import React from 'react';
import {
  FormGenerator,
  getDefaultValues,
  getFieldOptions,
  buildLookupTable,
  getSectionErrors,
  evalCond,
  injectGenProps,
  isFieldFilled,
  isFieldEmpty,
  isFieldVisible,
  mapFieldChildren,
  isNilOrEmpty
} from '@isobar-us/redux-form-gen';

const mergeCustomOptions = (customOptions, options) => ({
  ...options,
  customFieldTypes: {...customOptions.customFieldTypes, ...(options.customFieldTypes || {})},
  customOperators: {...customOptions.customOperators, ...(options.customOperators || {})}
});

const wrapComponent = (Wrapped, customOptions) => {
  const ComponentWrapper = (options) => <Wrapped {...mergeCustomOptions(customOptions, options)} />;

  return ComponentWrapper;
};

const wrapFunction = (wrapped, customOptions) => {
  return (options, ...args) => wrapped(mergeCustomOptions(customOptions, options), ...args);
};

const createFormGenerator = (customOptions) => {
  return {
    // functions
    getDefaultValues: wrapFunction(getDefaultValues, customOptions),
    getFieldOptions: wrapFunction(getFieldOptions, customOptions),
    buildLookupTable: wrapFunction(buildLookupTable, customOptions),
    getSectionErrors: wrapFunction(getSectionErrors, customOptions),
    evalCond: wrapFunction(evalCond, customOptions),
    isFieldFilled: wrapFunction(isFieldFilled, customOptions),
    isFieldEmpty: wrapFunction(isFieldEmpty, customOptions),
    isFieldVisible: wrapFunction(isFieldVisible, customOptions),
    mapFieldChildren: wrapFunction(mapFieldChildren, customOptions),
    // components
    FormGenerator: wrapComponent(FormGenerator, customOptions),
    injectGenProps: (Form) => {
      const GenForm = injectGenProps(Form);
      const InjectedGenProps = (options) => <GenForm {...mergeCustomOptions(options)} />;
      return InjectedGenProps;
    },
    // non-wrapped
    isNilOrEmpty
  };
};

export default createFormGenerator;
