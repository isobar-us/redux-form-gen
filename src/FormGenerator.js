// @flow
import PropTypes from 'prop-types';
import React, {Component} from 'react';
import {consumeReduxFormContext, genContext} from './contextUtils';
import Frag from './Frag';

import GenField from './GenField';

import omit from 'lodash/omit';
import set from 'lodash/set';
import get from 'lodash/get';
import cloneDeep from 'lodash/cloneDeep';

import type {Props, State} from './FormGenerator.types';

/**
 * [FormGenerator description]
 * @extends Component
 */
class FormGenerator extends Component<Props, State> {
  static childContextTypes = genContext;
  getChildContext = () => {
    return {
      gen: {
        ...omit(this.props, '_reduxForm'),
        getCachedValue: this.getCachedValue,
        setCachedValue: this.setCachedValue
      }
    };
  };

  static propTypes = {
    fields: PropTypes.array.isRequired, // an array of field objects
    customQuestionProps: PropTypes.object, // object key = questionId, value = an object of props for the specified questionId's
    visibleDepth: PropTypes.string, // dot-notation string, used to rendering fields after this path (startsWith)
    customFieldTypes: PropTypes.objectOf(PropTypes.func),
    display: PropTypes.oneOf(['stacked', 'inline']),
    disabled: PropTypes.bool // if true, disables all fields in the form
  };

  static defaultProps = {
    display: 'stacked'
  };

  state = {
    cachedValues: {} // TODO find a better way to do value caching for hidden fields
  };

  setCachedValue = (questionId, value) => {
    this.setState((prevState) => ({
      ...prevState,
      cachedValues: set(cloneDeep(prevState.cachedValues), questionId, value)
    }));
  };

  getCachedValue = (questionId) => {
    return get(this.state.cachedValues, questionId);
  };

  clearCachedValues = () => {
    this.setState({
      cachedValues: {}
    });
  };

  componentWillMount = () => {
    if (!this.props._reduxForm) {
      throw new Error('FormGenerator must be inside a component decorated with reduxForm()');
    }
  };

  render() {
    let {fields, children} = this.props;
    let path = 'fields';

    return fields ? (
      <Frag>
        <div className='generated-form'>
          {fields.map((field, index) => <GenField key={index} {...{field, path: `${path}[${index}]`}} />)}
        </div>
        {children}
      </Frag>
    ) : null;
  }
}

export default consumeReduxFormContext(FormGenerator);
