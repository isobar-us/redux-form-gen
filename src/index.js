// @flow
import FormGenerator from './FormGenerator';

import defaultFieldTypes, {getFieldOptions, genericFieldProps} from './defaultFieldTypes';
import GenericRequiredLabel from './defaultFieldTypes/components/GenericRequiredLabel';
import RequiredIndicator from './defaultFieldTypes/components/RequiredIndicator';
import injectGenProps from './injectGenProps';

import {
  isSectionEmpty,
  isSectionEmptyIterator,
  isSectionFilled,
  isSectionFilledIterator,
  getSectionErrors,
  getSectionErrorsIterator,
  isNilOrEmpty,
  getFieldPath,
  isFieldHidden,
  isFieldVisible,
  isFieldRequired,
  isFieldDisabled,
  isFieldEmpty,
  isFieldFilled,
  isFieldValid,
  mapFieldChildren
} from './validators';

import {getDefaultValues, buildLookupTable} from './utils';

import {consumeGenContext, GenContext} from './contextUtils';

import {evalCond, evalCondValid} from './conditionalUtils';

import GenField from './GenField';
import GenWrapper from './GenWrapper';

export type {FieldOptions, CustomFieldTypes} from './types';

export {
  FormGenerator,
  // defaultFieldTypes
  defaultFieldTypes,
  getFieldOptions,
  RequiredIndicator,
  GenericRequiredLabel,
  genericFieldProps,
  injectGenProps,
  // validators
  isSectionEmpty,
  isSectionEmptyIterator,
  isSectionFilled,
  isSectionFilledIterator,
  getSectionErrors,
  getSectionErrorsIterator,
  isNilOrEmpty,
  // validator utils
  getFieldPath,
  isFieldHidden,
  isFieldVisible,
  isFieldRequired,
  isFieldDisabled,
  isFieldEmpty,
  isFieldFilled,
  isFieldValid,
  mapFieldChildren,
  // utils
  getDefaultValues,
  buildLookupTable,
  // contextUtils
  consumeGenContext,
  GenContext,
  // conditionalUtils
  evalCond,
  evalCondValid,
  // internals
  GenField,
  GenWrapper
};
