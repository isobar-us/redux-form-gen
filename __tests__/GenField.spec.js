import React from 'react';
import {mount} from 'enzyme';
import {GenField, FormGenerator, GenContext} from '../src';

import {reduxForm, reducer as formReducer} from 'redux-form';
import {createStore, combineReducers} from 'redux';
import {Provider} from 'react-redux';

const reducer = combineReducers({form: formReducer});
let store = createStore(reducer, {});

const Form = reduxForm({
  form: 'testForm'
})(({children}) => <div>{children}</div>);
// const isDirtySelector = isDirty('testForm');
// const getFormValuesSelector = getFormValues('testForm');

const FormGenDecorator = ({children, formProps, genProps}) => (
  <Provider store={store}>
    <Form {...formProps}>
      <FormGenerator fields={[]} {...genProps}>
        {children}
      </FormGenerator>
    </Form>
  </Provider>
);

// reset the redux store after each test
beforeEach(() => {
  store = createStore(reducer, {});
});

let consoleErrorSpy;
let consoleWarnSpy;
const startErrorSupression = () => {
  consoleErrorSpy = jest.spyOn(console, 'error');
  consoleErrorSpy.mockImplementation(() => {}); // suppress error log
  consoleWarnSpy = jest.spyOn(console, 'warn');
  consoleWarnSpy.mockImplementation(() => {}); // suppress warn log
};

const stopErrorSuppression = () => {
  consoleErrorSpy.mockReset();
  consoleErrorSpy.mockRestore();
  consoleWarnSpy.mockReset();
  consoleWarnSpy.mockRestore();
};

describe('GenField', () => {
  it('should throw an error if not rendered inside a FormGenerator', () => {
    startErrorSupression();
    expect(() => {
      mount(<GenField />);
    }).toThrowErrorMatchingSnapshot();
    stopErrorSuppression();
  });

  it('should log an error if attempting to render an unknown type', () => {
    startErrorSupression();
    mount(
      <FormGenDecorator>
        <GenField field={{type: 'foo'}} path='fields[0]' />
      </FormGenDecorator>
    );
    expect(consoleWarnSpy.mock.calls.length).toBe(0);
    expect(consoleErrorSpy.mock.calls.length).toBe(2);
    expect(consoleErrorSpy.mock.calls[0]).toMatchSnapshot(); // getFieldOptions error
    expect(consoleErrorSpy.mock.calls[1]).toMatchSnapshot(); // GenField error
    stopErrorSuppression();
  });

  it('should log an error if attempting to render a field that is missing a name', () => {
    startErrorSupression();
    mount(
      <FormGenDecorator>
        <GenField field={{type: 'text'}} path='fields[0]' />
      </FormGenDecorator>
    );
    expect(consoleWarnSpy.mock.calls.length).toBe(0);
    expect(consoleErrorSpy.mock.calls.length).toBe(1);
    expect(consoleErrorSpy.mock.calls[0]).toMatchSnapshot(); // getFieldOptions error
    stopErrorSuppression();
  });

  it('should log an error if path is missing', () => {
    startErrorSupression();
    mount(
      <FormGenDecorator>
        <GenField field={{type: 'text', questionId: 'foo'}} />
      </FormGenDecorator>
    );
    expect(consoleWarnSpy.mock.calls.length).toBe(0);
    expect(consoleErrorSpy.mock.calls.length).toBe(2);
    expect(consoleErrorSpy.mock.calls).toMatchSnapshot(); // path error
    stopErrorSuppression();
  });

  it('should warn backup check for filled', () => {
    const fooField = {
      type: 'text',
      questionId: 'foo',
      conditionalVisible: {
        questionId: 'bar'
      }
    };

    startErrorSupression();
    mount(
      <Provider store={store}>
        <Form>
          <GenContext.Provider value={{wasGenerated: true}}>
            <GenField field={fooField} path='fields[0]' />
          </GenContext.Provider>
        </Form>
      </Provider>
    );
    expect(consoleWarnSpy.mock.calls.length).toBe(1);
    expect(consoleErrorSpy.mock.calls.length).toBe(0);
    expect(consoleWarnSpy.mock.calls).toMatchSnapshot();
    stopErrorSuppression();
  });

  it('should use the field type def for _genIsFilled', () => {
    const customFieldTypes = {
      bar: () => ({
        _genIsFilled: () => true
      })
    };
    const fooField = {
      type: 'text',
      questionId: 'foo',
      conditionalVisible: {
        questionId: 'bar'
      }
    };

    const barField = {
      type: 'bar',
      questionId: 'bar'
    };
    const wrapper = mount(
      <FormGenDecorator
        genProps={{
          lookupTable: {
            bar: barField
          },
          customFieldTypes
        }}
      >
        <GenField field={fooField} path='fields[0]' />
      </FormGenDecorator>
    );

    expect(wrapper.find('.section--hidden').length).toBe(0);
  });

  it('should use backup check for filled', () => {
    const customFieldTypes = {
      bar: () => ({
        _genIsFilled: () => true
      })
    };

    const fooField = {
      type: 'text',
      questionId: 'foo',
      conditionalVisible: {
        questionId: 'bar'
      }
    };

    const wrapper = mount(
      <FormGenDecorator
        genProps={{
          customFieldTypes
        }}
      >
        <GenField field={fooField} path='fields[0]' />
      </FormGenDecorator>
    );

    expect(wrapper.find('.section--hidden').length).toBe(1);
  });
});
