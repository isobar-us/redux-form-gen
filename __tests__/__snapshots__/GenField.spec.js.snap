// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenField should log an error if attempting to render a field that is missing a name 1`] = `
Array [
  "Form Generator: you must specify a questionId for this field",
  Object {
    "type": "text",
  },
]
`;

exports[`GenField should log an error if attempting to render an unknown type 1`] = `
Array [
  "Form Generator: unknown field type \\"foo\\". 
Field:",
  Object {
    "type": "foo",
  },
  "
Known Types:",
  Array [
    "text",
    "textarea",
    "select",
    "radio",
    "group",
    "section",
    "array",
    "arrayItem",
  ],
]
`;

exports[`GenField should log an error if attempting to render an unknown type 2`] = `
Array [
  "Form Generator: unknown field type \\"foo\\". 
Field:",
  Object {
    "type": "foo",
  },
  "
. skipping render.",
]
`;

exports[`GenField should log an error if path is missing 1`] = `
Array [
  Array [
    "Missing path for ",
    Object {
      "questionId": "foo",
      "type": "text",
    },
  ],
  Array [
    "Missing path for ",
    Object {
      "questionId": "foo",
      "type": "text",
    },
  ],
]
`;

exports[`GenField should throw an error if not rendered inside a FormGenerator 1`] = `"GenField must be rendered as a child of a <FormGenerator>"`;

exports[`GenField should warn backup check for filled 1`] = `
Array [
  Array [
    "[Form Generator] attempted to check \`filled\` without a lookupTable. Condition:",
    Object {
      "questionId": "bar",
    },
  ],
]
`;
